File: ecmascript/triangle/triangle.js
export default class Triangle {
  constructor(a, b, c) {
    this.a = a;
    this.b = b;
    this.c = c;
    this.sides = [a,b,c];
  }
  kind() {
    this.checkInequality();
    if (this.sides.filter((side) => side === this.a).length === 3) {return 'equilateral';}
    if (this.sides.filter((side) => side === this.a).length === 2 ||
        this.sides.filter((side) => side === this.b).length === 2) {return 'isosceles';}
    return 'scalene';
  }
  checkInequality(){
    if ((this.a <= 0) || (this.b <= 0) || (this.c <= 0)) {
      throw new Error();
      return;
    }
    if (this.a + this.b < this.c) {throw new Error(); return; }
    if (this.b + this.c < this.a) {throw new Error(); return; }
    if (this.c + this.a < this.b) {throw new Error(); return; }
  }
}

File: elixir/triangle/triangle.exs
defmodule Validity do
  defmacro are_all_positive(a, b, c), do: quote do: unquote(c) > 0 and unquote(a) > 0 and unquote(b) > 0
  defmacro is_valid(a, b, c), do: quote do: unquote(c) < (unquote(a) + unquote(b))
end

defmodule Triangle do
  import Validity

  def kind(a, b, c) do
    [aa, bb, cc] = Enum.sort([a, b, c])
    get_kind(aa, bb, cc)
  end

  defp get_kind(a, b, c) when not are_all_positive(a, b, c), do: {:error, "all side lengths must be positive"}
  defp get_kind(a, b, c) when not is_valid(a, b, c), do: {:error, "side lengths violate triangle inequality"}

  defp get_kind(a, a, a), do: {:ok, :equilateral}
  defp get_kind(_, a, a), do: {:ok, :isosceles}
  defp get_kind(a, _, a), do: {:ok, :isosceles}
  defp get_kind(a, a, _), do: {:ok, :isosceles}
  defp get_kind(_, _, _), do: {:ok, :scalene}
end
